logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/debug/RepairFaultRemovalKeysUseCase.kt

```diff
@@ -0,0 +1,163 @@
+) : RepairFaultRemovalKeysUseCase {
+
+    val logger by lazy { kaliumLogger.withTextTag("RepairFaultRemovalKeysUseCase") }
+
+    override suspend fun invoke(param: TargetedRepairParam): RepairResult = withContext(dispatcher.io) {
+        if (selfUserId.domain != param.domain) {
+            logger.w(
+                "Attempted to repair faulty removal keys in domain ${param.domain}, but user belongs to ${selfUserId.domain}. Aborting."
+            )
+            return@withContext RepairResult.RepairNotNeeded
+        }
+
+        conversationRepository.getMLSConversationsByDomain(param.domain)
+            .flatMap { conversations ->
+                val conversationsWithFaultyKeys = conversations.filter { convo ->
+                    checkConversationHasFaultyKey(convo.protocol, param.faultyKey)
+                }
+
+                when {
+                    conversationsWithFaultyKeys.isEmpty() -> RepairResult.NoConversationsToRepair.right()
+                    else -> repairConversations(conversationsWithFaultyKeys, conversations.size).right()
+                }
+            }
+            .fold(
+                fnL = {
+                    logger.e("Error occurred during repair of faulty removal keys")
+                    RepairResult.Error
+                },
+                fnR = { it }
+            )
+    }
+
+    private suspend fun repairConversations(
+        conversationsToRepair: List<Conversation>,
+        totalChecked: Int
+    ): RepairResult {
+        val (successful, failed) = conversationsToRepair
+            .map { convo -> convo to delegateResetMLSConversation(convo.id) }
+            .partition { (_, result) -> result.isRight() }
+
+        return RepairResult.RepairPerformed(
+            totalConversationsChecked = totalChecked,
+            conversationsWithFaultyKeys = conversationsToRepair.size,
+            successfullyRepairedConversations = successful.size,
+            failedRepairs = failed.map { (convo, _) -> convo.id.toLogString() }
+        )
+    }
+
+    @Suppress("TooGenericExceptionCaught")
+    private suspend fun delegateResetMLSConversation(conversationId: ConversationId): Either<CoreFailure, Unit> = try {
+        transactionProvider.transaction("RepairFaultRemovalKeys") { context ->
+            resetMLSConversation(conversationId, context)
+        }
+    } catch (exception: Exception) {
+        logger.e("Exception during resetting MLS conversation ${conversationId.toLogString()}", exception)
+        CoreFailure.Unknown(exception).left()
+    }
+
+    private suspend fun checkConversationHasFaultyKey(protocolInfo: Conversation.ProtocolInfo, faultyKey: String): Boolean =
+        when (protocolInfo) {
+            Conversation.ProtocolInfo.Proteus -> false
+            is Conversation.ProtocolInfo.MLS,
+            is Conversation.ProtocolInfo.Mixed -> {
+                transactionProvider.mlsTransaction<Boolean>("CheckFaultyRemovalKey") { context ->
```

Comment:
Same here, open one transaction to entire use case and use here:
`transactionContext.wrapInMLSContext{}`